In [1]:
import pandas as pd
import numpy as np
import nsepy
from datetime import date
from bokeh.plotting import figure 
from bokeh.io import output_notebook, show
output_notebook()

Loading BokehJS ...

### Part  3 : Data Modelling
Collect latest data (from 1st Jan 2015 to 5th April 2019) to train the model to predict tomorrow's data.

In [2]:
tcs = nsepy.get_history(symbol='TCS', start=date(2015,1,1), end=date(2019,4,5))
infy = nsepy.get_history(symbol='INFY', start=date(2015,1,1), end=date(2019,4,5))

In [3]:
tcs

Symbol Series  Prev Close     Open     High      Low     Last  \
Date                                                                       
2015-01-01    TCS     EQ     2558.25  2567.00  2567.00  2541.00  2550.00   
2015-01-02    TCS     EQ     2545.55  2551.00  2590.95  2550.60  2588.40   
2015-01-05    TCS     EQ     2579.45  2581.00  2599.90  2524.65  2538.10   
2015-01-06    TCS     EQ     2540.25  2529.10  2529.10  2440.00  2450.05   
2015-01-07    TCS     EQ     2446.60  2470.00  2479.15  2407.45  2426.90   
2015-01-08    TCS     EQ     2417.70  2442.40  2449.00  2420.55  2446.00   
2015-01-09    TCS     EQ     2443.80  2455.00  2519.90  2450.00  2510.00   
2015-01-12    TCS     EQ     2512.30  2517.00  2528.00  2480.25  2527.95   
2015-01-13    TCS     EQ     2509.70  2520.00  2530.40  2480.10  2498.00   
2015-01-14    TCS     EQ     2497.90  2516.00  2531.80  2500.50  2525.05   
2015-01-15    TCS     EQ     2521.95  2533.00  2581.75  2523.25  2542.05   
2015-01-16    TCS     EQ     2539.10  2543.70  2545.00  2484.30  2526.00   
2015-01-19    TCS     EQ     2532.15  2536.00  2551.00  2503.00  2509.00   
2015-01-20    TCS     EQ     2511.00  2520.00  2530.00  2483.25  2498.00   
2015-01-21    TCS     EQ     2500.45  2522.00  2522.00  2488.25  2518.00   
2015-01-22    TCS     EQ     2513.80  2516.50  2521.25  2495.50  2511.00   
2015-01-23    TCS     EQ     2513.55  2514.00  2532.05  2496.10  2504.10   
2015-01-27    TCS     EQ     2503.60  2512.35  2515.00  2489.55  2504.20   
2015-01-28    TCS     EQ     2502.05  2503.00  2544.90  2502.90  2533.40   
2015-01-29    TCS     EQ     2539.60  2538.00  2568.70  2512.90  2560.70   
2015-01-30    TCS     EQ     2544.65  2560.70  2575.00  2470.00  2484.00   
2015-02-02    TCS     EQ     2482.05  2482.00  2521.45  2462.05  2516.00   
2015-02-03    TCS     EQ     2514.20  2511.65  2565.00  2507.10  2552.25   
2015-02-04    TCS     EQ     2558.25  2568.00  2571.70  2502.10  2514.95   
2015-02-05    TCS     EQ     2514.30  2520.00  2591.70  2519.00  2553.00   
2015-02-06    TCS     EQ     2552.40  2553.00  2594.80  2550.85  2576.10   
2015-02-09    TCS     EQ     2575.75  2561.50  2569.10  2500.00  2512.00   
2015-02-10    TCS     EQ     2512.90  2510.00  2526.25  2433.00  2445.50   
2015-02-11    TCS     EQ     2441.15  2455.05  2487.25  2447.25  2460.00   
2015-02-12    TCS     EQ     2459.90  2479.80  2479.80  2447.00  2469.00   
...           ...    ...         ...      ...      ...      ...      ...   
2019-02-21    TCS     EQ     1914.75  1914.75  1940.00  1897.05  1912.00   
2019-02-22    TCS     EQ     1914.20  1917.20  1930.00  1905.10  1929.00   
2019-02-25    TCS     EQ     1925.65  1932.50  1990.00  1930.50  1987.00   
2019-02-26    TCS     EQ     1985.15  1984.00  2045.15  1972.25  2032.50   
2019-02-27    TCS     EQ     2038.70  2040.00  2074.95  2022.00  2052.50   
2019-02-28    TCS     EQ     2058.10  2060.00  2071.35  1977.60  1983.50   
2019-03-01    TCS     EQ     1983.45  1995.05  2005.00  1980.35  1995.00   
2019-03-05    TCS     EQ     1995.40  2005.00  2007.00  1976.60  1985.05   
2019-03-06    TCS     EQ     1988.10  1989.30  2015.00  1985.05  2005.00   
2019-03-07    TCS     EQ     1999.60  2005.00  2024.05  2000.20  2015.00   
2019-03-08    TCS     EQ     2013.30  2025.00  2033.00  2010.05  2022.75   
2019-03-11    TCS     EQ     2022.70  2028.90  2033.00  2003.65  2016.15   
2019-03-12    TCS     EQ     2014.80  2014.05  2024.80  2003.00  2009.10   
2019-03-13    TCS     EQ     2012.45  2013.00  2015.90  1978.60  1995.00   
2019-03-14    TCS     EQ     2000.50  2004.95  2007.80  1981.00  1990.40   
2019-03-15    TCS     EQ     1987.40  1998.90  2068.95  1991.00  2036.00   
2019-03-18    TCS     EQ     2039.95  2043.00  2064.60  2011.00  2023.85   
2019-03-19    TCS     EQ     2022.80  2030.00  2030.00  1995.10  2028.50   
2019-03-20    TCS     EQ     2022.80  2028.00  2044.80  2000.00  2000.25   
2019-03-22    TCS     EQ     2015.05  2015.00  2016.00  198

In [5]:
tcs.corr()

Prev Close      Open      High       Low      Last  \
Prev Close            1.000000  0.985356  0.983367  0.983519  0.980877   
Open                  0.985356  1.000000  0.997514  0.997803  0.995065   
High                  0.983367  0.997514  1.000000  0.997611  0.998177   
Low                   0.983519  0.997803  0.997611  1.000000  0.997975   
Last                  0.980877  0.995065  0.998177  0.997975  1.000000   
Close                 0.980811  0.995139  0.998270  0.998096  0.999884   
VWAP                  0.982866  0.997135  0.999194  0.999074  0.999180   
Volume               -0.176289 -0.193402 -0.180936 -0.201585 -0.192423   
Turnover             -0.003769 -0.017334 -0.004939 -0.024763 -0.016063   
Trades               -0.297202 -0.311531 -0.295849 -0.324950 -0.312277   
Deliverable Volume   -0.138082 -0.158329 -0.151739 -0.165633 -0.160928   
%Deliverble           0.043657  0.042321  0.024663  0.046701  0.032751   

                       Close      VWAP    Volume  Turnover    Trades  \
Prev Close          0.980811  0.982866 -0.176289 -0.003769 -0.297202   
Open                0.995139  0.997135 -0.193402 -0.017334 -0.311531   
High                0.998270  0.999194 -0.180936 -0.004939 -0.295849   
Low                 0.998096  0.999074 -0.201585 -0.024763 -0.324950   
Last                0.999884  0.999180 -0.192423 -0.016063 -0.312277   
Close               1.000000  0.999315 -0.191922 -0.015577 -0.311812   
VWAP                0.999315  1.000000 -0.192641 -0.016186 -0.310662   
Volume             -0.191922 -0.192641  1.000000  0.970411  0.684261   
Turnover           -0.015577 -0.016186  0.970411  1.000000  0.578101   
Trades             -0.311812 -0.310662  0.684261  0.578101  1.000000   
Deliverable Volume -0.160642 -0.161012  0.951945  0.955481  0.552466   
%Deliverble         0.032467  0.033357 -0.160700 -0.129952 -0.221578   

                    Deliverable Volume  %Deliverble  
Prev Close                   -0.138082     0.043657  
Open                         -0.158329     0.042321  
High                         -0.151739     0.024663  
Low                          -0.165633     0.046701  
Last                         -0.160928     0.032751  
Close                        -0.160642     0.032467  
VWAP                         -0.161012     0.033357  
Volume                        0.951945    -0.160700  
Turnover                      0.955481    -0.129952  
Trades                        0.552466    -0.221578  
Deliverable Volume            1.000000     0.043382  
%Deliverble                   0.043382     1.000000

In [4]:
infy

Symbol Series  Prev Close     Open     High      Low     Last  \
Date                                                                       
2015-01-01   INFY     EQ     1972.55  1968.95  1982.00  1956.90  1971.00   
2015-01-02   INFY     EQ     1974.40  1972.00  2019.05  1972.00  2017.95   
2015-01-05   INFY     EQ     2013.20  2009.90  2030.00  1977.50  1996.00   
2015-01-06   INFY     EQ     1995.90  1980.00  1985.00  1934.10  1965.10   
2015-01-07   INFY     EQ     1954.20  1965.00  1974.75  1950.00  1966.05   
2015-01-08   INFY     EQ     1963.55  1985.60  1997.00  1950.00  1979.25   
2015-01-09   INFY     EQ     1973.45  1980.10  2109.00  1913.05  2075.30   
2015-01-12   INFY     EQ     2074.45  2092.00  2119.20  2075.00  2112.95   
2015-01-13   INFY     EQ     2115.95  2107.80  2107.80  2075.00  2092.00   
2015-01-14   INFY     EQ     2088.90  2098.50  2133.00  2092.60  2129.00   
2015-01-15   INFY     EQ     2128.65  2135.20  2152.80  2112.25  2130.00   
2015-01-16   INFY     EQ     2134.25  2148.00  2160.00  2112.00  2115.00   
2015-01-19   INFY     EQ     2118.30  2120.00  2132.35  2094.60  2105.00   
2015-01-20   INFY     EQ     2102.50  2107.00  2133.00  2101.45  2119.20   
2015-01-21   INFY     EQ     2124.20  2131.50  2175.95  2118.00  2172.00   
2015-01-22   INFY     EQ     2169.90  2174.80  2203.00  2161.30  2194.95   
2015-01-23   INFY     EQ     2196.45  2200.00  2225.00  2165.10  2211.90   
2015-01-27   INFY     EQ     2215.05  2221.00  2221.00  2130.10  2137.55   
2015-01-28   INFY     EQ     2136.50  2150.00  2154.80  2130.30  2144.45   
2015-01-29   INFY     EQ     2144.55  2145.00  2149.90  2127.60  2147.80   
2015-01-30   INFY     EQ     2145.50  2161.50  2162.55  2126.95  2141.85   
2015-02-02   INFY     EQ     2141.90  2141.50  2152.00  2126.05  2137.00   
2015-02-03   INFY     EQ     2136.85  2136.85  2139.95  2102.00  2124.00   
2015-02-04   INFY     EQ     2120.90  2139.90  2150.00  2126.40  2143.80   
2015-02-05   INFY     EQ     2142.85  2150.80  2220.00  2148.00  2186.00   
2015-02-06   INFY     EQ     2193.80  2205.00  2247.80  2188.50  2228.00   
2015-02-09   INFY     EQ     2230.50  2200.00  2254.00  2200.00  2245.00   
2015-02-10   INFY     EQ     2248.90  2243.00  2298.00  2241.00  2268.00   
2015-02-11   INFY     EQ     2278.30  2280.50  2293.20  2256.25  2287.80   
2015-02-12   INFY     EQ     2284.85  2283.00  2321.95  2273.00  2309.90   
...           ...    ...         ...      ...      ...      ...      ...   
2019-02-21   INFY     EQ      740.70   738.00   738.00   727.10   734.00   
2019-02-22   INFY     EQ      733.45   732.50   738.20   728.00   734.90   
2019-02-25   INFY     EQ      734.95   735.55   758.00   735.55   755.00   
2019-02-26   INFY     EQ      754.90   748.60   752.60   740.15   741.55   
2019-02-27   INFY     EQ      742.50   745.00   746.40   732.75   734.50   
2019-02-28   INFY     EQ      735.25   738.70   742.40   732.30   734.70   
2019-03-01   INFY     EQ      734.30   735.50   743.40   735.50   741.80   
2019-03-05   INFY     EQ      741.90   740.50   742.50   730.25   732.75   
2019-03-06   INFY     EQ      732.50   736.10   736.10   727.00   732.25   
2019-03-07   INFY     EQ      732.50   734.00   734.50   720.55   723.30   
2019-03-08   INFY     EQ      722.95   722.95   722.95   710.65   713.20   
2019-03-11   INFY     EQ      712.35   713.30   718.50   709.40   711.05   
2019-03-12   INFY     EQ      711.25   709.00   714.95   705.60   706.75   
2019-03-13   INFY     EQ      706.95   710.00   718.60   706.30   709.00   
2019-03-14   INFY     EQ      708.45   714.00   714.90   706.20   708.45   
2019-03-15   INFY     EQ      708.35   712.00   723.75   709.60   720.00   
2019-03-18   INFY     EQ      718.55   721.55   723.70   705.35   709.50   
2019-03-19   INFY     EQ      710.20   712.80   726.50   707.40   725.90   
2019-03-20   INFY     EQ      722.25   727.25   742.30   727.15   737.80   
2019-03-22   INFY     EQ      738.95   745.00   746.35   73

In [6]:
infy.corr()

Prev Close      Open      High       Low      Last  \
Prev Close            1.000000  0.994816  0.994539  0.994297  0.994095   
Open                  0.994816  1.000000  0.999509  0.999388  0.999010   
High                  0.994539  0.999509  1.000000  0.999299  0.999534   
Low                   0.994297  0.999388  0.999299  1.000000  0.999610   
Last                  0.994095  0.999010  0.999534  0.999610  1.000000   
Close                 0.994075  0.999000  0.999537  0.999623  0.999979   
VWAP                  0.994446  0.999376  0.999756  0.999754  0.999856   
Volume               -0.292666 -0.294580 -0.289261 -0.307868 -0.300685   
Turnover             -0.029635 -0.029191 -0.021921 -0.044335 -0.035384   
Trades               -0.224693 -0.225964 -0.220158 -0.241075 -0.233472   
Deliverable Volume   -0.307873 -0.308459 -0.304359 -0.319928 -0.313874   
%Deliverble           0.199624  0.204937  0.197814  0.211050  0.204495   

                       Close      VWAP    Volume  Turnover    Trades  \
Prev Close          0.994075  0.994446 -0.292666 -0.029635 -0.224693   
Open                0.999000  0.999376 -0.294580 -0.029191 -0.225964   
High                0.999537  0.999756 -0.289261 -0.021921 -0.220158   
Low                 0.999623  0.999754 -0.307868 -0.044335 -0.241075   
Last                0.999979  0.999856 -0.300685 -0.035384 -0.233472   
Close               1.000000  0.999876 -0.300757 -0.035455 -0.233428   
VWAP                0.999876  1.000000 -0.299592 -0.033889 -0.231797   
Volume             -0.300757 -0.299592  1.000000  0.935934  0.902788   
Turnover           -0.035455 -0.033889  0.935934  1.000000  0.881732   
Trades             -0.233428 -0.231797  0.902788  0.881732  1.000000   
Deliverable Volume -0.313995 -0.313237  0.961634  0.894626  0.847186   
%Deliverble         0.204456  0.203864 -0.270861 -0.212571 -0.280951   

                    Deliverable Volume  %Deliverble  
Prev Close                   -0.307873     0.199624  
Open                         -0.308459     0.204937  
High                         -0.304359     0.197814  
Low                          -0.319928     0.211050  
Last                         -0.313874     0.204495  
Close                        -0.313995     0.204456  
VWAP                         -0.313237     0.203864  
Volume                        0.961634    -0.270861  
Turnover                      0.894626    -0.212571  
Trades                        0.847186    -0.280951  
Deliverable Volume            1.000000    -0.060175  
%Deliverble                  -0.060175     1.000000

Feature engineering is done to extract features (last 7 days' closing price and a rolling average calculation of window 10) to predict the close price.

Also, the data is divided into training and test sets to check the model accuracy.

In [7]:
tcs_day = [date.day for date in tcs.index.values]
tcs_month = [date.month for date in tcs.index.values]
tcs_prev_close_1 = tcs['Prev Close'].values.reshape(1, -1)
tcs_prev_close_2 = tcs.Close.shift(2)
tcs_prev_close_2 = tcs_prev_close_2.fillna(0)
tcs_prev_close_3 = tcs.Close.shift(3)
tcs_prev_close_3 = tcs_prev_close_3.fillna(0)
tcs_prev_close_4 = tcs.Close.shift(4)
tcs_prev_close_4 = tcs_prev_close_4.fillna(0)
tcs_prev_close_5 = tcs.Close.shift(5)
tcs_prev_close_5 = tcs_prev_close_5.fillna(0)
tcs_prev_close_6 = tcs.Close.shift(6)
tcs_prev_close_6 = tcs_prev_close_6.fillna(0)
tcs_prev_close_7 = tcs.Close.shift(7)
tcs_prev_close_7 = tcs_prev_close_7.fillna(0)
tcs_roll_avg_10 = tcs['Close'].rolling(10).mean()
tcs_roll_avg_10 = tcs_roll_avg_10.fillna(0)
tcs_X = np.dstack((tcs_day, tcs_month, tcs_prev_close_1, tcs_prev_close_2, tcs_prev_close_3, tcs_prev_close_4, tcs_prev_close_5, tcs_prev_close_6, tcs_prev_close_7, tcs_roll_avg_10))[0]
tcs_y = tcs.Close.values
tcs_X_train = tcs_X[0:-20,:]
tcs_y_train = tcs_y[0:-20]
tcs_X_test = tcs_X[-20:,:]
tcs_y_test = tcs_y[-20:]

In [8]:
print(tcs_X_train.shape)
print(tcs_y_train.shape)
print(tcs_X_test.shape)
print(tcs_y_test.shape)

(1036, 10)
(1036,)
(20, 10)
(20,)


In [9]:
p = figure(title= 'TCS Close Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Close Price', plot_width=750, plot_height=350)
p.line(tcs.index, tcs.Close, legend="Close", line_width=2)
p.line(tcs.index, tcs_prev_close_1[0], legend="tcs_prev_close_1", line_color='red', line_width=2, alpha=0.4)
p.line(tcs.index, tcs_prev_close_2, legend="tcs_prev_close_2", line_color='brown', line_width=2, alpha=0.4)
p.line(tcs.index, tcs_prev_close_3, legend="tcs_prev_close_3", line_color='purple', line_width=2, alpha=0.4)
p.line(tcs.index, tcs_prev_close_4, legend="tcs_prev_close_4", line_color='pink', line_width=2, alpha=0.4)
p.line(tcs.index, tcs_prev_close_5, legend="tcs_prev_close_5", line_color='green', line_width=2, alpha=0.4)
p.line(tcs.index, tcs_prev_close_6, legend="tcs_prev_close_6", line_color='yellow', line_width=2, alpha=0.4)
p.line(tcs.index, tcs_prev_close_7, legend="tcs_prev_close_7", line_color='orange', line_width=2, alpha=0.4)
p.line(tcs.index, tcs_roll_avg_10, legend="tcs_roll_avg_10", line_color='black', line_width=2, alpha=0.4)
p.legend.background_fill_alpha = 0.3
p.legend.label_text_font_size = "5pt"
p.legend.label_height = 0
p.legend.glyph_height = 15
p.legend.glyph_width = 25
p.legend.padding = 5
p.legend.spacing = 0
p.legend.margin = 0
show(p)

In [10]:
infy_day = [date.day for date in infy.index.values]
infy_month = [date.month for date in infy.index.values]
infy_prev_close_1 = infy['Prev Close'].values.reshape(1, -1)
infy_prev_close_2 = infy.Close.shift(2)
infy_prev_close_2 = infy_prev_close_2.fillna(0)
infy_prev_close_3 = infy.Close.shift(3)
infy_prev_close_3 = infy_prev_close_3.fillna(0)
infy_prev_close_4 = infy.Close.shift(4)
infy_prev_close_4 = infy_prev_close_4.fillna(0)
infy_prev_close_5 = infy.Close.shift(5)
infy_prev_close_5 = infy_prev_close_5.fillna(0)
infy_prev_close_6 = infy.Close.shift(6)
infy_prev_close_6 = infy_prev_close_6.fillna(0)
infy_prev_close_7 = infy.Close.shift(7)
infy_prev_close_7 = infy_prev_close_7.fillna(0)
infy_roll_avg_10 = infy['Close'].rolling(10).mean()
infy_roll_avg_10 = infy_roll_avg_10.fillna(0)
infy_X = np.dstack((infy_day, infy_month, infy_prev_close_1, infy_prev_close_2, infy_prev_close_3, infy_prev_close_4, infy_prev_close_5, infy_prev_close_6, infy_prev_close_7, infy_roll_avg_10))[0]
infy_y = infy.Close.values
infy_X_train = infy_X[0:-20,:]
infy_y_train = infy_y[0:-20]
infy_X_test = infy_X[-20:,:]
infy_y_test = infy_y[-20:]

In [11]:
print(infy_X_train.shape)
print(infy_y_train.shape)
print(infy_X_test.shape)
print(infy_y_test.shape)

(1036, 10)
(1036,)
(20, 10)
(20,)


In [12]:
p = figure(title= 'INFY Close Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Close Price', plot_width=750, plot_height=350)
p.line(infy.index, infy.Close, legend="Close", line_width=2)
p.line(infy.index, infy_prev_close_1[0], legend="infy_prev_close_1", line_color='red', line_width=2, alpha=0.4)
p.line(infy.index, infy_prev_close_2, legend="infy_prev_close_2", line_color='brown', line_width=2, alpha=0.4)
p.line(infy.index, infy_prev_close_3, legend="infy_prev_close_3", line_color='purple', line_width=2, alpha=0.4)
p.line(infy.index, infy_prev_close_4, legend="infy_prev_close_4", line_color='pink', line_width=2, alpha=0.4)
p.line(infy.index, infy_prev_close_5, legend="infy_prev_close_5", line_color='green', line_width=2, alpha=0.4)
p.line(infy.index, infy_prev_close_6, legend="infy_prev_close_6", line_color='yellow', line_width=2, alpha=0.4)
p.line(infy.index, infy_prev_close_7, legend="infy_prev_close_7", line_color='orange', line_width=2, alpha=0.4)
p.line(infy.index, infy_roll_avg_10, legend="infy_roll_avg_10", line_color='black', line_width=2, alpha=0.4)
p.legend.background_fill_alpha = 0.3
p.legend.label_text_font_size = "5pt"
p.legend.label_height = 0
p.legend.glyph_height = 15
p.legend.glyph_width = 25
p.legend.padding = 5
p.legend.spacing = 0
p.legend.margin = 0
show(p)

##### 1. Quick build any 2 Models
I took the 5 model examples given in the question and tried a default model of each to check which of those models perform better.

In [13]:
from sklearn.svm import SVR
from sklearn.linear_model import LinearRegression, Ridge, LassoLars
from sklearn.ensemble import GradientBoostingRegressor

In [14]:
from time import time

I had to modify the existing mean_absolute_error function of scikit-learn, so that I could calulate the mean_absolute_percentage_error (MAPE) as required by this task.

In [15]:
# https://github.com/scikit-learn/scikit-learn/blob/7b136e9/sklearn/metrics/regression.py#L112

from sklearn.utils import check_array, check_consistent_length
from sklearn.externals.six import string_types

def _check_reg_targets(y_true, y_pred, multioutput):
    check_consistent_length(y_true, y_pred)
    y_true = check_array(y_true, ensure_2d=False)
    y_pred = check_array(y_pred, ensure_2d=False)

    if y_true.ndim == 1:
        y_true = y_true.reshape((-1, 1))

    if y_pred.ndim == 1:
        y_pred = y_pred.reshape((-1, 1))

    if y_true.shape[1] != y_pred.shape[1]:
        raise ValueError("y_true and y_pred have different number of output "
                         "({0}!={1})".format(y_true.shape[1], y_pred.shape[1]))

    n_outputs = y_true.shape[1]
    allowed_multioutput_str = ('raw_values', 'uniform_average',
                               'variance_weighted')
    if isinstance(multioutput, string_types):
        if multioutput not in allowed_multioutput_str:
            raise ValueError("Allowed 'multioutput' string values are {}. "
                             "You provided multioutput={!r}".format(
                                 allowed_multioutput_str,
                                 multioutput))
    elif multioutput is not None:
        multioutput = check_array(multioutput, ensure_2d=False)
        if n_outputs == 1:
            raise ValueError("Custom weights are useful only in "
                             "multi-output cases.")
        elif n_outputs != len(multioutput):
            raise ValueError(("There must be equally many custom weights "
                              "(%d) as outputs (%d).") %
                             (len(multioutput), n_outputs))
    y_type = 'continuous' if n_outputs == 1 else 'continuous-multioutput'

    return y_type, y_true, y_pred, multioutput

def mean_absolute_percentage_error(y_true, y_pred):
    y_type, y_true, y_pred, multioutput = _check_reg_targets(y_true, y_pred, multioutput='uniform_average')
    check_consistent_length(y_true, y_pred, None)
    output_errors = np.average(np.abs((y_true - y_pred) / y_true), axis=0)
    

    return (np.average(output_errors)*100)

While training the default models for TCS data, we also take note of the prediction time, total time (prediction time + training time), training MAPE, test MAPE, and the plot of the actual and predicted data.

In [16]:
lr = LinearRegression()
svr = SVR()
r = Ridge()
ll = LassoLars()
gbr = GradientBoostingRegressor()

In [17]:
start_time_1 = time()
lr.fit(tcs_X_train, tcs_y_train)
start_time_2 = time()
tcs_y_tr_pred = lr.predict(tcs_X_train)
tcs_y_te_pred = lr.predict(tcs_X_test)
end_time = time()
print("Prediction Time --- %s ms ---" % ((end_time - start_time_2)*1000))
print("Total Time --- %s ms ---" % ((end_time - start_time_1)*1000))
print("TCS Train MAPE Linear Regression: " + str(mean_absolute_percentage_error(tcs_y_train, tcs_y_tr_pred)) + " %")
print("TCS Test MAPE Linear Regression: " + str(mean_absolute_percentage_error(tcs_y_test, tcs_y_te_pred)) + " %")

tcs_y_pred = np.hstack([tcs_y_tr_pred, tcs_y_te_pred])
p = figure(title= 'TCS Linear Regression Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Close Price', plot_width=750, plot_height=350)
p.line(tcs.index, tcs.Close, legend="Close", line_width=2)
p.line(tcs.index, tcs_y_pred, legend="tcs_y_pred", line_color='red', line_width=2, alpha=0.4)
show(p)

Prediction Time --- 0.3635883331298828 ms ---
Total Time --- 25.307416915893555 ms ---
TCS Train MAPE Linear Regression: 1.1458078647864234 %
TCS Test MAPE Linear Regression: 1.1144986491306759 %


In [18]:
start_time_1 = time()
svr.fit(tcs_X_train, tcs_y_train)
start_time_2 = time()
tcs_y_tr_pred = svr.predict(tcs_X_train)
tcs_y_te_pred = svr.predict(tcs_X_test)
end_time = time()
print("Prediction Time --- %s ms ---" % ((end_time - start_time_2)*1000))
print("Total Time --- %s ms ---" % ((end_time - start_time_1)*1000))
print("TCS Train MAPE SVR: " + str(mean_absolute_percentage_error(tcs_y_train, tcs_y_tr_pred)) + " %")
print("TCS Test MAPE SVR: " + str(mean_absolute_percentage_error(tcs_y_test, tcs_y_te_pred)) + " %")

tcs_y_pred = np.hstack([tcs_y_tr_pred, tcs_y_te_pred])
p = figure(title= 'TCS SVR Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Close Price', plot_width=750, plot_height=350)
p.line(tcs.index, tcs.Close, legend="Close", line_width=2)
p.line(tcs.index, tcs_y_pred, legend="tcs_y_pred", line_color='red', line_width=2, alpha=0.4)
show(p)

Prediction Time --- 24.93906021118164 ms ---
Total Time --- 67.56758689880371 ms ---
TCS Train MAPE SVR: 9.854094526461704 %
TCS Test MAPE SVR: 22.747720593563876 %


/home/saswat/anaconda3/envs/my_env/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


In [19]:
start_time_1 = time()
r.fit(tcs_X_train, tcs_y_train)
start_time_2 = time()
tcs_y_tr_pred = r.predict(tcs_X_train)
tcs_y_te_pred = r.predict(tcs_X_test)
end_time = time()
print("Prediction Time --- %s ms ---" % ((end_time - start_time_2)*1000))
print("Total Time --- %s ms ---" % ((end_time - start_time_1)*1000))
print("TCS Train MAPE Ridge Regression: " + str(mean_absolute_percentage_error(tcs_y_train, tcs_y_tr_pred)) + " %")
print("TCS Test MAPE Ridge Regression: " + str(mean_absolute_percentage_error(tcs_y_test, tcs_y_te_pred)) + " %")

tcs_y_pred = np.hstack([tcs_y_tr_pred, tcs_y_te_pred])
p = figure(title= 'TCS Ridge Regression Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Close Price', plot_width=750, plot_height=350)
p.line(tcs.index, tcs.Close, legend="Close", line_width=2)
p.line(tcs.index, tcs_y_pred, legend="tcs_y_pred", line_color='red', line_width=2, alpha=0.4)
show(p)

Prediction Time --- 0.4227161407470703 ms ---
Total Time --- 2.260923385620117 ms ---
TCS Train MAPE Ridge Regression: 1.1458076189367303 %
TCS Test MAPE Ridge Regression: 1.1144941806790736 %


In [20]:
start_time_1 = time()
ll.fit(tcs_X_train, tcs_y_train)
start_time_2 = time()
tcs_y_tr_pred = ll.predict(tcs_X_train)
tcs_y_te_pred = ll.predict(tcs_X_test)
end_time = time()
print("Prediction Time --- %s ms ---" % ((end_time - start_time_2)*1000))
print("Total Time --- %s ms ---" % ((end_time - start_time_1)*1000))
print("TCS Train MAPE LassoLars: " + str(mean_absolute_percentage_error(tcs_y_train, tcs_y_tr_pred)) + " %")
print("TCS Test MAPE LassoLars: " + str(mean_absolute_percentage_error(tcs_y_test, tcs_y_te_pred)) + " %")

tcs_y_pred = np.hstack([tcs_y_tr_pred, tcs_y_te_pred])
p = figure(title= 'TCS LassoLars Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Close Price', plot_width=750, plot_height=350)
p.line(tcs.index, tcs.Close, legend="Close", line_width=2)
p.line(tcs.index, tcs_y_pred, legend="tcs_y_pred", line_color='red', line_width=2, alpha=0.4)
show(p)

Prediction Time --- 0.37384033203125 ms ---
Total Time --- 3.5486221313476562 ms ---
TCS Train MAPE LassoLars: 1.6663274144589884 %
TCS Test MAPE LassoLars: 2.38128657581376 %


In [21]:
start_time_1 = time()
gbr.fit(tcs_X_train, tcs_y_train)
start_time_2 = time()
tcs_y_tr_pred = gbr.predict(tcs_X_train)
tcs_y_te_pred = gbr.predict(tcs_X_test)
end_time = time()
print("Prediction Time --- %s ms ---" % ((end_time - start_time_2)*1000))
print("Total Time --- %s ms ---" % ((end_time - start_time_1)*1000))
print("TCS Train MAPE GB Regressor: " + str(mean_absolute_percentage_error(tcs_y_train, tcs_y_tr_pred)) + " %")
print("TCS Test MAPE GB Regressor: " + str(mean_absolute_percentage_error(tcs_y_test, tcs_y_te_pred)) + " %")

tcs_y_pred = np.hstack([tcs_y_tr_pred, tcs_y_te_pred])
p = figure(title= 'TCS GB Regressor Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Close Price', plot_width=750, plot_height=350)
p.line(tcs.index, tcs.Close, legend="Close", line_width=2)
p.line(tcs.index, tcs_y_pred, legend="tcs_y_pred", line_color='red', line_width=2, alpha=0.4)
show(p)

Prediction Time --- 1.0592937469482422 ms ---
Total Time --- 89.46728706359863 ms ---
TCS Train MAPE GB Regressor: 0.7984450393017835 %
TCS Test MAPE GB Regressor: 0.9308819248961316 %


While training the default models for the INFY data, we also take note of the prediction time, total time (prediction time + training time), training MAPE, test MAPE, and the plot of the actual and predicted data.

In [22]:
lr = LinearRegression()
svr = SVR()
r = Ridge()
ll = LassoLars()
gbr = GradientBoostingRegressor()

In [23]:
start_time_1 = time()
lr.fit(infy_X_train, infy_y_train)
start_time_2 = time()
infy_y_tr_pred = lr.predict(infy_X_train)
infy_y_te_pred = lr.predict(infy_X_test)
end_time = time()
print("Prediction Time --- %s ms ---" % ((end_time - start_time_2)*1000))
print("Total Time --- %s ms ---" % ((end_time - start_time_1)*1000))
print("INFY Train MAPE Linear Regression: " + str(mean_absolute_percentage_error(infy_y_train, infy_y_tr_pred)) + " %")
print("INFY Test MAPE Linear Regression: " + str(mean_absolute_percentage_error(infy_y_test, infy_y_te_pred)) + " %")

infy_y_pred = np.hstack([infy_y_tr_pred, infy_y_te_pred])
p = figure(title= 'INFY Linear Regression Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Close Price', plot_width=750, plot_height=350)
p.line(infy.index, infy.Close, legend="Close", line_width=2)
p.line(infy.index, infy_y_pred, legend="infy_y_pred", line_color='red', line_width=2, alpha=0.4)
show(p)

Prediction Time --- 0.2651214599609375 ms ---
Total Time --- 4.476308822631836 ms ---
INFY Train MAPE Linear Regression: 1.3326741250158973 %
INFY Test MAPE Linear Regression: 1.0100084906279054 %


In [24]:
start_time_1 = time()
svr.fit(infy_X_train, infy_y_train)
start_time_2 = time()
infy_y_tr_pred = svr.predict(infy_X_train)
infy_y_te_pred = svr.predict(infy_X_test)
end_time = time()
print("Prediction Time --- %s ms ---" % ((end_time - start_time_2)*1000))
print("Total Time --- %s ms ---" % ((end_time - start_time_1)*1000))
print("INFY Train MAPE SVR: " + str(mean_absolute_percentage_error(infy_y_train, infy_y_tr_pred)) + " %")
print("INFY Test MAPE SVR: " + str(mean_absolute_percentage_error(infy_y_test, infy_y_te_pred)) + " %")

infy_y_pred = np.hstack([infy_y_tr_pred, infy_y_te_pred])
p = figure(title= 'INFY SVR Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Close Price', plot_width=750, plot_height=350)
p.line(infy.index, infy.Close, legend="Close", line_width=2)
p.line(infy.index, infy_y_pred, legend="infy_y_pred", line_color='red', line_width=2, alpha=0.4)
show(p)

Prediction Time --- 33.05673599243164 ms ---
Total Time --- 78.4611701965332 ms ---
INFY Train MAPE SVR: 18.39517793247386 %
INFY Test MAPE SVR: 44.95055665640444 %


/home/saswat/anaconda3/envs/my_env/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


In [25]:
start_time_1 = time()
r.fit(infy_X_train, infy_y_train)
start_time_2 = time()
infy_y_tr_pred = r.predict(infy_X_train)
infy_y_te_pred = r.predict(infy_X_test)
end_time = time()
print("Prediction Time --- %s ms ---" % ((end_time - start_time_2)*1000))
print("Total Time --- %s ms ---" % ((end_time - start_time_1)*1000))
print("INFY Train MAPE Ridge Regression: " + str(mean_absolute_percentage_error(infy_y_train, infy_y_tr_pred)) + " %")
print("INFY Test MAPE Ridge Regression: " + str(mean_absolute_percentage_error(infy_y_test, infy_y_te_pred)) + " %")

infy_y_pred = np.hstack([infy_y_tr_pred, infy_y_te_pred])
p = figure(title= 'INFY Ridge Regression Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Close Price', plot_width=750, plot_height=350)
p.line(infy.index, infy.Close, legend="Close", line_width=2)
p.line(infy.index, infy_y_pred, legend="infy_y_pred", line_color='red', line_width=2, alpha=0.4)
show(p)

Prediction Time --- 0.3209114074707031 ms ---
Total Time --- 3.133058547973633 ms ---
INFY Train MAPE Ridge Regression: 1.3326713573766702 %
INFY Test MAPE Ridge Regression: 1.009997219010343 %


In [26]:
start_time_1 = time()
ll.fit(infy_X_train, infy_y_train)
start_time_2 = time()
infy_y_tr_pred = ll.predict(infy_X_train)
infy_y_te_pred = ll.predict(infy_X_test)
end_time = time()
print("Prediction Time --- %s ms ---" % ((end_time - start_time_2)*1000))
print("Total Time --- %s ms ---" % ((end_time - start_time_1)*1000))
print("INFY Train MAPE LassoLars: " + str(mean_absolute_percentage_error(infy_y_train, infy_y_tr_pred)) + " %")
print("INFY Test MAPE LassoLars: " + str(mean_absolute_percentage_error(infy_y_test, infy_y_te_pred)) + " %")

infy_y_pred = np.hstack([infy_y_tr_pred, infy_y_te_pred])
p = figure(title= 'INFY LassoLars Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Close Price', plot_width=750, plot_height=350)
p.line(infy.index, infy.Close, legend="Close", line_width=2)
p.line(infy.index, infy_y_pred, legend="infy_y_pred", line_color='red', line_width=2, alpha=0.4)
show(p)

Prediction Time --- 0.385284423828125 ms ---
Total Time --- 2.407073974609375 ms ---
INFY Train MAPE LassoLars: 2.4770847949120554 %
INFY Test MAPE LassoLars: 4.885371450263384 %


In [27]:
start_time_1 = time()
gbr.fit(infy_X_train, infy_y_train)
start_time_2 = time()
infy_y_tr_pred = gbr.predict(infy_X_train)
infy_y_te_pred = gbr.predict(infy_X_test)
end_time = time()
print("Prediction Time --- %s ms ---" % ((end_time - start_time_2)*1000))
print("Total Time --- %s ms ---" % ((end_time - start_time_1)*1000))
print("INFY Train MAPE GB Regressor: " + str(mean_absolute_percentage_error(infy_y_train, infy_y_tr_pred)) + " %")
print("INFY Test MAPE GB Regressor: " + str(mean_absolute_percentage_error(infy_y_test, infy_y_te_pred)) + " %")

infy_y_pred = np.hstack([infy_y_tr_pred, infy_y_te_pred])
p = figure(title= 'INFY GB Regressor Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Close Price', plot_width=750, plot_height=350)
p.line(infy.index, infy.Close, legend="Close", line_width=2)
p.line(infy.index, infy_y_pred, legend="infy_y_pred", line_color='red', line_width=2, alpha=0.4)
show(p)

Prediction Time --- 1.0075569152832031 ms ---
Total Time --- 89.7221565246582 ms ---
INFY Train MAPE GB Regressor: 0.9283590492923768 %
INFY Test MAPE GB Regressor: 1.1579758453243372 %


OBSERVATIONS:

From above, we observe that Ridge Regression and Gradient Boosting Regressor performs better as compared to others.

But, if we also consider the time taken for prediction, then Ridge Regression is slightly better.

##### 2. Test using pytest
I am unable to understand the questions. I know that pytest is generally used to verify whether a particular function is working properly or not. But, in this code there is no such function to verify (using testcases).

To produce the test data (validation data) for the model, I have taken the last 20 days data from the TCS dataframe, and last 10 days data from the INFY dataframe. This has been done before starting the quick build process.

##### 3. Model Assumptions

As Ridge Regression (included in sklearn.linear_model) is similar to Linear Regression. Most of the assumptions of Linear Regression (OLS Regression) will also be applicable to Ridge Regression.

From https://www.albert.io/blog/key-assumptions-of-ols-econometrics-review/, we can see some assumptions of OLS Regression.

1. Assumption-1 and Assumption-5 may hold true (upto a certain extent), as in the model we are supplying data for the past 7 days' price and also the rolling average calculation with a window of 10.

2. Assumption-2 should hold true, because Ridge Regression considers each data point (dependent variable) as unique, and only related to its corresponding independent values (in which we have added the past few days' data).

3. Assumption-3 should be true because it is natural for the model to sometimes predict higher values (than the actual) and sometimes lower values. So, the mean of errors should be close to zero.

4. Assumption-4 will hold true as there is only one independent variable.

5. Please refer to Point-1 above.

6. Assumption-6 may or may not hold true, but as said in the post this assumption is not compulsory for the validity of the model.

##### 4. Tuning the Best Model

Let us tune the Ridge Regression model for both TCS and INFY independently, and check whether the training and testing MAPE can be improved (reduced).

In [28]:
r_tcs = Ridge(alpha = 100000.0, fit_intercept=False, random_state = 42)

In [29]:
start_time_1 = time()
r_tcs.fit(tcs_X_train, tcs_y_train)
start_time_2 = time()
tcs_y_tr_pred = r_tcs.predict(tcs_X_train)
tcs_y_te_pred = r_tcs.predict(tcs_X_test)
end_time = time()
print("Prediction Time --- %s ms ---" % ((end_time - start_time_2)*1000))
print("Total Time --- %s ms ---" % ((end_time - start_time_1)*1000))
print("TCS Train MAPE Ridge Regression: " + str(mean_absolute_percentage_error(tcs_y_train, tcs_y_tr_pred)) + " %")
print("TCS Test MAPE Ridge Regression: " + str(mean_absolute_percentage_error(tcs_y_test, tcs_y_te_pred)) + " %")

tcs_y_pred = np.hstack([tcs_y_tr_pred, tcs_y_te_pred])
p = figure(title= 'TCS Ridge Regression Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Close Price', plot_width=750, plot_height=350)
p.line(tcs.index, tcs.Close, legend="Close", line_width=2)
p.line(tcs.index, tcs_y_pred, legend="tcs_y_pred", line_color='red', line_width=2, alpha=0.4)
show(p)

Prediction Time --- 0.4391670227050781 ms ---
Total Time --- 3.212451934814453 ms ---
TCS Train MAPE Ridge Regression: 1.138131279306045 %
TCS Test MAPE Ridge Regression: 0.9240872013951771 %


The MAPE reduced slightly for the TCS stock data.

Old Train MAPE: 1.1458076189367303 %,
Old Test MAPE: 1.1144941806790736 %

Tuned Train MAPE: 1.138131279306045 %,
Tuned Test MAPE: 0.9240872013951771 %

PLEASE NOTE:

Prediction Time --- 0.4391670227050781 ms ---
Total Time --- 3.212451934814453 ms ---

So, later (in the python_stock_predictor.py file) even if it is required to retrain the data again, then also it will not take much time (as Ridge Regression model gets trained very fast).

Please see Note at the end of this Jupyter Notebook.

In [30]:
r_infy = Ridge(alpha = 100000.0, fit_intercept = False, random_state = 42)

In [31]:
start_time_1 = time()
r_infy.fit(infy_X_train, infy_y_train)
start_time_2 = time()
infy_y_tr_pred = r_infy.predict(infy_X_train)
infy_y_te_pred = r_infy.predict(infy_X_test)
end_time = time()
print("Prediction Time --- %s ms ---" % ((end_time - start_time_2)*1000))
print("Total Time --- %s ms ---" % ((end_time - start_time_1)*1000))
print("INFY Train MAPE Ridge Regression: " + str(mean_absolute_percentage_error(infy_y_train, infy_y_tr_pred)) + " %")
print("INFY Test MAPE Ridge Regression: " + str(mean_absolute_percentage_error(infy_y_test, infy_y_te_pred)) + " %")

infy_y_pred = np.hstack([infy_y_tr_pred, infy_y_te_pred])
p = figure(title= 'INFY Ridge Regression Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Close Price', plot_width=750, plot_height=350)
p.line(infy.index, infy.Close, legend="Close", line_width=2)
p.line(infy.index, infy_y_pred, legend="infy_y_pred", line_color='red', line_width=2, alpha=0.4)
show(p)

Prediction Time --- 0.4487037658691406 ms ---
Total Time --- 3.490447998046875 ms ---
INFY Train MAPE Ridge Regression: 1.313966857475498 %
INFY Test MAPE Ridge Regression: 0.9229112943717261 %


The MAPE reduced slightly for INFY stock data.

Old Train MAPE: 1.3326713573766702 %,
Old Test MAPE: 1.009997219010343 %

Tuned Train MAPE: 1.313966857475498 %,
Tuned Test MAPE: 0.9229112943717261 %

PLEASE NOTE:

Prediction Time --- 0.4487037658691406 ms ---
Total Time --- 3.490447998046875 ms ---

So, later (in the python_stock_predictor.py file) even if it is required to retrain the data again, then also it will not take much time (as Ridge Regression model gets trained very fast).

Please see Note at the end of this Jupyter Notebook.

##### 5. Final MAPE of the Models

Tuned TCS Train MAPE: 1.138131279306045 %, Tuned TCS Test MAPE: 0.9240872013951771 %

Tuned INFY Train MAPE: 1.313966857475498 %, Tuned INFY Test MAPE: 0.9229112943717261 %

##### 6. Model to Predict Volume Shock

Feature engineering is done to extract required features (last 7 days' volume and a rolling average calculation of window 10) to predict tomorrow's volume shock.

Also, the data is divided into training and test sets to check the model accuracy.

In [32]:
tcs_day = [date.day for date in tcs.index.values]
tcs_month = [date.month for date in tcs.index.values]
tcs_prev_volume_1 = tcs.Volume.shift(1)
tcs_vol_y = (abs((tcs["Volume"] - tcs_prev_volume_1)/tcs["Volume"]) > 0.1).astype(int)
tcs_vol_y = np.asarray(tcs_vol_y)

tcs_prev_volume_1 = tcs_prev_volume_1.fillna(0)
tcs_prev_volume_2 = tcs.Volume.shift(2)
tcs_prev_volume_2 = tcs_prev_volume_2.fillna(0)
tcs_prev_volume_3 = tcs.Volume.shift(3)
tcs_prev_volume_3 = tcs_prev_volume_3.fillna(0)
tcs_prev_volume_4 = tcs.Volume.shift(4)
tcs_prev_volume_4 = tcs_prev_volume_4.fillna(0)
tcs_prev_volume_5 = tcs.Volume.shift(5)
tcs_prev_volume_5 = tcs_prev_volume_5.fillna(0)
tcs_prev_volume_6 = tcs.Volume.shift(6)
tcs_prev_volume_6 = tcs_prev_volume_6.fillna(0)
tcs_prev_volume_7 = tcs.Volume.shift(7)
tcs_prev_volume_7 = tcs_prev_volume_7.fillna(0)
tcs_vol_roll_avg_10 = tcs['Volume'].rolling(10).mean()
tcs_vol_roll_avg_10 = tcs_vol_roll_avg_10.fillna(0)
tcs_vol_X = np.dstack((tcs_day, tcs_month, tcs_prev_volume_1, tcs_prev_volume_2, tcs_prev_volume_3, tcs_prev_volume_4, tcs_prev_volume_5, tcs_prev_volume_6, tcs_prev_volume_7, tcs_vol_roll_avg_10))[0]

In [33]:
tcs_vol_X_train = tcs_vol_X[0:-20,:]
tcs_vol_y_train = tcs_vol_y[0:-20]
tcs_vol_X_test = tcs_vol_X[-20:,:]
tcs_vol_y_test = tcs_vol_y[-20:]
print(tcs_vol_X_train.shape)
print(tcs_vol_y_train.shape)
print(tcs_vol_X_test.shape)
print(tcs_vol_y_test.shape)

(1036, 10)
(1036,)
(20, 10)
(20,)


In [34]:
p = figure(title= 'TCS Volume Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='', plot_width=750, plot_height=350)
p.circle(tcs.index, tcs_vol_y*10000000, legend="Volume Shock", size=0.2)
p.line(tcs.index, tcs_prev_volume_1, legend="tcs_prev_volume_1", line_color='red', line_width=2, alpha=0.3)
p.line(tcs.index, tcs_prev_volume_2, legend="tcs_prev_volume_2", line_color='brown', line_width=2, alpha=0.3)
p.line(tcs.index, tcs_prev_volume_3, legend="tcs_prev_volume_3", line_color='purple', line_width=2, alpha=0.3)
p.line(tcs.index, tcs_prev_volume_4, legend="tcs_prev_volume_4", line_color='pink', line_width=2, alpha=0.3)
p.line(tcs.index, tcs_prev_volume_5, legend="tcs_prev_volume_5", line_color='green', line_width=2, alpha=0.3)
p.line(tcs.index, tcs_prev_volume_6, legend="tcs_prev_volume_6", line_color='yellow', line_width=2, alpha=0.3)
p.line(tcs.index, tcs_prev_volume_7, legend="tcs_prev_volume_7", line_color='orange', line_width=2, alpha=0.3)
p.line(tcs.index, tcs_vol_roll_avg_10, legend="tcs_vol_roll_avg_10", line_color='black', line_width=2, alpha=0.3)
p.legend.background_fill_alpha = 0.3
p.legend.label_text_font_size = "5pt"
p.legend.label_height = 0
p.legend.glyph_height = 15
p.legend.glyph_width = 25
p.legend.padding = 5
p.legend.spacing = 0
p.legend.margin = 0
show(p)

In [35]:
infy_day = [date.day for date in infy.index.values]
infy_month = [date.month for date in infy.index.values]
infy_prev_volume_1 = infy.Volume.shift(1)
infy_vol_y = (abs((infy["Volume"] - infy_prev_volume_1)/infy["Volume"]) > 0.1).astype(int)
infy_vol_y = np.asarray(infy_vol_y)

infy_prev_volume_1 = infy_prev_volume_1.fillna(0)
infy_prev_volume_2 = infy.Volume.shift(2)
infy_prev_volume_2 = infy_prev_volume_2.fillna(0)
infy_prev_volume_3 = infy.Volume.shift(3)
infy_prev_volume_3 = infy_prev_volume_3.fillna(0)
infy_prev_volume_4 = infy.Volume.shift(4)
infy_prev_volume_4 = infy_prev_volume_4.fillna(0)
infy_prev_volume_5 = infy.Volume.shift(5)
infy_prev_volume_5 = infy_prev_volume_5.fillna(0)
infy_prev_volume_6 = infy.Volume.shift(6)
infy_prev_volume_6 = infy_prev_volume_6.fillna(0)
infy_prev_volume_7 = infy.Volume.shift(7)
infy_prev_volume_7 = infy_prev_volume_7.fillna(0)
infy_vol_roll_avg_10 = infy['Volume'].rolling(10).mean()
infy_vol_roll_avg_10 = infy_vol_roll_avg_10.fillna(0)
infy_vol_X = np.dstack((infy_day, infy_month, infy_prev_volume_1, infy_prev_volume_2, infy_prev_volume_3, infy_prev_volume_4, infy_prev_volume_5, infy_prev_volume_6, infy_prev_volume_7, infy_vol_roll_avg_10))[0]

In [36]:
infy_vol_X_train = infy_vol_X[0:-20,:]
infy_vol_y_train = infy_vol_y[0:-20]
infy_vol_X_test = infy_vol_X[-20:,:]
infy_vol_y_test = infy_vol_y[-20:]
print(infy_vol_X_train.shape)
print(infy_vol_y_train.shape)
print(infy_vol_X_test.shape)
print(infy_vol_y_test.shape)

(1036, 10)
(1036,)
(20, 10)
(20,)


In [37]:
p = figure(title= 'INFY Volume Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='', plot_width=750, plot_height=350)
p.circle(infy.index, infy_vol_y*10000000, legend="Volume Shock", size=0.2)
p.line(infy.index, infy_prev_volume_1, legend="infy_prev_volume_1", line_color='red', line_width=2, alpha=0.3)
p.line(infy.index, infy_prev_volume_2, legend="infy_prev_volume_2", line_color='brown', line_width=2, alpha=0.3)
p.line(infy.index, infy_prev_volume_3, legend="infy_prev_volume_3", line_color='purple', line_width=2, alpha=0.3)
p.line(infy.index, infy_prev_volume_4, legend="infy_prev_volume_4", line_color='pink', line_width=2, alpha=0.3)
p.line(infy.index, infy_prev_volume_5, legend="infy_prev_volume_5", line_color='green', line_width=2, alpha=0.3)
p.line(infy.index, infy_prev_volume_6, legend="infy_prev_volume_6", line_color='yellow', line_width=2, alpha=0.3)
p.line(infy.index, infy_prev_volume_7, legend="infy_prev_volume_7", line_color='orange', line_width=2, alpha=0.3)
p.line(infy.index, infy_vol_roll_avg_10, legend="infy_vol_roll_avg_10", line_color='black', line_width=2, alpha=0.3)
p.legend.background_fill_alpha = 0.3
p.legend.label_text_font_size = "5pt"
p.legend.label_height = 0
p.legend.glyph_height = 15
p.legend.glyph_width = 25
p.legend.padding = 5
p.legend.spacing = 0
p.legend.margin = 0
show(p)

In [38]:
# As volume shock is binary data, we will use accuracy_score and GradientBoostingClassifier

from sklearn.metrics import accuracy_score, f1_score
from sklearn.ensemble import GradientBoostingClassifier

In [39]:
gbc_tcs_vol = GradientBoostingClassifier(n_estimators = 300, max_depth = 2, max_features = None, random_state = 2139)

In [40]:
start_time_1 = time()
gbc_tcs_vol.fit(tcs_vol_X_train, tcs_vol_y_train)
start_time_2 = time()
tcs_vol_y_tr_pred = gbc_tcs_vol.predict(tcs_vol_X_train)
tcs_vol_y_te_pred = gbc_tcs_vol.predict(tcs_vol_X_test)
end_time = time()
print("Prediction Time --- %s ms ---" % ((end_time - start_time_2)*1000))
print("Total Time --- %s ms ---" % ((end_time - start_time_1)*1000))
print("TCS Vol Shock Train Accuracy GBC: " + str(accuracy_score(tcs_vol_y_train, tcs_vol_y_tr_pred)))
print("TCS Vol Shock Test Accuracy GBC: " + str(accuracy_score(tcs_vol_y_test, tcs_vol_y_te_pred)))
print("TCS Vol Shock Train F1-Score GBC: " + str(f1_score(tcs_vol_y_train, tcs_vol_y_tr_pred)))
print("TCS Vol Shock Test F1-Score GBC: " + str(f1_score(tcs_vol_y_test, tcs_vol_y_te_pred)))

tcs_vol_y_pred = np.hstack([tcs_vol_y_tr_pred, tcs_vol_y_te_pred])
p = figure(title= 'TCS Vol Shock GBC Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Volume Shock', plot_width=750, plot_height=350)
p.circle(tcs.index, tcs_vol_y, legend="tcs_vol_y", size=0.2)
p.circle(tcs.index, tcs_vol_y_pred, legend="tcs_vol_y_pred", color='red', size=0.2, alpha=0.4)
show(p)

Prediction Time --- 2.4166107177734375 ms ---
Total Time --- 242.93017387390137 ms ---
TCS Vol Shock Train Accuracy GBC: 0.9044401544401545
TCS Vol Shock Test Accuracy GBC: 0.75
TCS Vol Shock Train F1-Score GBC: 0.9449694274597
TCS Vol Shock Test F1-Score GBC: 0.8571428571428571


In [41]:
gbc_infy_vol = GradientBoostingClassifier(n_estimators = 200, max_depth = 3, max_features = None, random_state = 2139)

In [42]:
start_time_1 = time()
gbc_infy_vol.fit(infy_vol_X_train, infy_vol_y_train)
start_time_2 = time()
infy_vol_y_tr_pred = gbc_infy_vol.predict(infy_vol_X_train)
infy_vol_y_te_pred = gbc_infy_vol.predict(infy_vol_X_test)
end_time = time()
print("Prediction Time --- %s ms ---" % ((end_time - start_time_2)*1000))
print("Total Time --- %s ms ---" % ((end_time - start_time_1)*1000))
print("INFY Vol Shock Train Accuracy GBC: " + str(accuracy_score(infy_vol_y_train, infy_vol_y_tr_pred)))
print("INFY Vol Shock Test Accuracy GBC: " + str(accuracy_score(infy_vol_y_test, infy_vol_y_te_pred)))
print("INFY Vol Shock Train F1-Score GBC: " + str(f1_score(infy_vol_y_train, infy_vol_y_tr_pred)))
print("INFY Vol Shock Test F1-Score GBC: " + str(f1_score(infy_vol_y_test, infy_vol_y_te_pred)))

infy_vol_y_pred = np.hstack([infy_vol_y_tr_pred, infy_vol_y_te_pred])
p = figure(title= 'INFY Vol Shock GBC Plot', x_axis_label='Date', x_axis_type="datetime", y_axis_label='Volume Shock', plot_width=750, plot_height=350)
p.circle(infy.index, infy_vol_y, legend="infy_vol_y", size=0.2)
p.circle(infy.index, infy_vol_y_pred, legend="infy_vol_y_pred", color='red', size=0.2, alpha=0.4)
show(p)

Prediction Time --- 2.369403839111328 ms ---
Total Time --- 234.68756675720215 ms ---
INFY Vol Shock Train Accuracy GBC: 0.9411196911196911
INFY Vol Shock Test Accuracy GBC: 0.75
INFY Vol Shock Train F1-Score GBC: 0.965749578888265
INFY Vol Shock Test F1-Score GBC: 0.8571428571428571


##### 7. Predict using Python File
We will create a file named python_stock_predictor.py which can be used directly through the Terminal to predict tomorrow's price.

As the above 4 models have been trained using the data from around past 4 years, we would like to store the model in its trained condition, so that we can directly predict tomorrow's data (in less than 100 ms - as required by the task). To do that, we will use the dump function of the sklearn.externals.joblib module.

In [43]:
from sklearn.externals import joblib

In [44]:
# Save the 'r_tcs' model (TCS Closing Price Model) to disk
joblib.dump(r_tcs, 'r_tcs_model.sav')

['r_tcs_model.sav']

In [45]:
# Save the 'r_infy' (INFY Closing Price Model) model to disk
joblib.dump(r_infy, 'r_infy_model.sav')

['r_infy_model.sav']

In [46]:
# Save the 'gbc_tcs_vol' (TCS Volume Shock Model) model to disk
joblib.dump(gbc_tcs_vol, 'gbc_tcs_vol_model.sav')

['gbc_tcs_vol_model.sav']

In [47]:
# Save the 'gbc_infy_vol' (INFY Volume Shock Model) model to disk
joblib.dump(gbc_infy_vol, 'gbc_infy_vol_model.sav')

['gbc_infy_vol_model.sav']

OBSERVATIONS:

When we use the python_stock_predictor.py file (which uses the pretrained model to make a prediction), we observe that the Prediction Time is 16.02 ms (for TCS) and 15.04 ms (for INFY). The Total Time (which is dependent on the internet speed) is 729 ms (for TCS) and 854 ms (for INFY).

So, our Prediction Time is less than 100 ms (as required).

PLEASE NOTE:

Prediction Time is the time taken for extracting the price features & volume features, and then using those features to make a prediction. (It is not dependent on internet speed)

Total Time is the sum of Prediction Time and the time taken to fetch the data from the NSE servers using nsepy. (It is dependent on internet speed and ping)

In [62]:
p = figure(x_range=(0,730), y_range=(-450,0),plot_width=730, plot_height=450)
p.image_url(url=['prediction_time.png'], x=None, y=None, h=450, w=730)
show(p)

PLEASE NOTE:

If it is required to train the data every time we execute the python file, then the code in the python file can be slightly modified to train on new data every time.

Still then, I expect that the Prediction Time (excluding the time taken to fetch the required data from the NSE servers) will be less than 100 ms, because as we noted earlier the total time taken for training the Ridge Regression model was also very less (around 2 ms).